File: server/src/routes/export.ts
import { Router } from 'express';
import { getSession } from '../services/sessionService.js';
import { exportToCSV, exportToJSON, exportSummaryToJSON } from '../services/exportService.js';
import type { SessionSource } from '../parsers/index.js';

export const exportRouter = Router();

// GET /api/export/:source/:sessionId - Export session data
exportRouter.get('/:source/:sessionId', (req, res) => {
  try {
    const { source, sessionId } = req.params;
    const format = (req.query.format as string) || 'json';

    if (source !== 'claude' && source !== 'copilot') {
      res.status(400).json({ error: 'Bad request', message: 'Invalid source. Must be "claude" or "copilot"' });
      return;
    }

    const session = getSession(source as SessionSource, sessionId);

    if (!session) {
      res.status(404).json({ error: 'Not found', message: 'Session not found' });
      return;
    }

    const filename = `${source}-${sessionId}`;

    switch (format) {
      case 'csv':
        res.setHeader('Content-Type', 'text/csv');
        res.setHeader('Content-Disposition', `attachment; filename="${filename}.csv"`);
        res.send(exportToCSV(session));
        break;

      case 'json':
        res.setHeader('Content-Type', 'application/json');
        res.setHeader('Content-Disposition', `attachment; filename="${filename}.json"`);
        res.send(exportToJSON(session));
        break;

      case 'summary':
        res.setHeader('Content-Type', 'application/json');
        res.setHeader('Content-Disposition', `attachment; filename="${filename}-summary.json"`);
        res.send(exportSummaryToJSON(session));
        break;

      default:
        res.status(400).json({
          error: 'Bad request',
          message: 'Invalid format. Must be "csv", "json", or "summary"',
        });
    }
  } catch (error) {
    console.error('Error exporting session:', error);
    res.status(500).json({ error: 'Internal server error', message: 'Failed to export session' });
  }
});

File: server-go/handlers/export.go
package handlers

import (
	"fmt"
	"net/http"

	"github.com/go-chi/chi/v5"

	"agent-session-viewer/services"
)

// ExportHandler binds /api/export routes.
func ExportHandler(r chi.Router) {
	r.Get("/{source}/{sessionId}", exportSession)
}

func exportSession(w http.ResponseWriter, r *http.Request) {
	src, ok := parseSource(w, chi.URLParam(r, "source"))
	if !ok {
		return
	}
	sessionID := chi.URLParam(r, "sessionId")
	format := r.URL.Query().Get("format")
	if format == "" {
		format = "json"
	}

	session := services.GetSession(src, sessionID)
	if session == nil {
		http.Error(w, `{"error":"Not found"}`, http.StatusNotFound)
		return
	}

	filename := fmt.Sprintf("%s-%s", src, sessionID)

	switch format {
	case "json":
		data, err := services.ExportToJSON(session)
		if err != nil {
			http.Error(w, `{"error":"Export failed"}`, http.StatusInternalServerError)
			return
		}
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s.json"`, filename))
		w.Write([]byte(data))
	case "csv":
		data := services.ExportToCSV(session)
		w.Header().Set("Content-Type", "text/csv")
		w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s.csv"`, filename))
		w.Write([]byte(data))
	case "summary":
		data, err := services.ExportSummaryToJSON(session)
		if err != nil {
			http.Error(w, `{"error":"Export failed"}`, http.StatusInternalServerError)
			return
		}
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s-summary.json"`, filename))
		w.Write([]byte(data))
	default:
		http.Error(w, `{"error":"Bad request","message":"Invalid format. Must be json, csv, or summary"}`, http.StatusBadRequest)
	}
}
